docs/source/basics/tokenizers.rst

```diff
@@ -0,0 +1,239 @@
+
+            Returns:
+                str: The decoded text.
+            """
+            pass
+
+If you load any :class:`~torchtune.modules.tokenizers.ModelTokenizer`, you can see that it calls its underlying :class:`~torchtune.modules.tokenizers.BaseTokenizer`
+to do the actual encoding and decoding.
+
+.. code-block:: python
+
+    from torchtune.models.mistral import mistral_tokenizer
+    from torchtune.modules.tokenizers import SentencePieceBaseTokenizer
+
+    m_tokenizer = mistral_tokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+    # Mistral uses SentencePiece for its underlying BPE
+    sp_tokenizer = SentencePieceBaseTokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+
+    text = "hello world"
+
+    print(m_tokenizer.encode(text))
+    # [1, 6312, 28709, 1526, 2]
+
+    print(sp_tokenizer.encode(text))
+    # [1, 6312, 28709, 1526, 2]
+
+
+Model tokenizers
+----------------
+
+:class:`~torchtune.modules.tokenizers.ModelTokenizer` are specific to a particular model. They are required to implement the ``tokenize_messages`` method,
+which converts a list of Messages into a list of token IDs.
+
+.. code-block:: python
+
+    class ModelTokenizer(Protocol):
+
+        special_tokens: Dict[str, int]
+        max_seq_len: Optional[int]
+
+        def tokenize_messages(
+            self, messages: List[Message], **kwargs: Dict[str, Any]
+        ) -> Tuple[List[int], List[bool]]:
+            """
+            Given a list of messages, return a list of tokens and list of masks for
+            the concatenated and formatted messages.
+
+            Args:
+                messages (List[Message]): The list of messages to tokenize.
+                **kwargs (Dict[str, Any]): kwargs.
+
+            Returns:
+                Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+            """
+            pass
+
+The reason they are model specific and different from :class:`~torchtune.modules.tokenizers.BaseTokenizer`
+is because they add all the necessary special tokens or prompt templates required to prompt the model.
+
+.. code-block:: python
+
+    from torchtune.models.mistral import mistral_tokenizer
+    from torchtune.modules.tokenizers import SentencePieceBaseTokenizer
+    from torchtune.data import Message
+
+    m_tokenizer = mistral_tokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+    # Mistral uses SentencePiece for its underlying BPE
+    sp_tokenizer = SentencePieceBaseTokenizer("/tmp/Mistral-7B-v0.1/tokenizer.model")
+
+    text = "hello world"
+    msg = Message(role="user", content=text)
+
+    tokens, mask = m_tokenizer.tokenize_messages([msg])
+    print(tokens)
+    # [1, 733, 16289, 28793, 6312, 28709, 1526, 28705, 733, 28748, 16289, 28793]
+    print(sp_tokenizer.encode(text))
+    # [1, 6312, 28709, 1526, 2]
+    print(m_tokenizer.decode(tokens))
+    # [INST] hello world  [/INST]
+    print(sp_tokenizer.decode(sp_tokenizer.encode(text)))
+    # hello world
```

Comment:
I think this is a very useful demonstration of the fact that the model tokenizer is handling the formatting and stitching together of messages with the appropriate special tokens. But just by looking through this snippet it's not really clear. I would break this down into more of a step-by-step walkthrough (point out which tokens are coming from user tag, which ones are special tokens, etc)